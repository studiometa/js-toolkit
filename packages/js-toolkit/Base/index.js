import { getComponentElements } from './utils.js';
import ChildrenManager from './managers/ChildrenManager.js';
import RefsManager from './managers/RefsManager.js';
import ServicesManager from './managers/ServicesManager.js';
import EventsManager from './managers/EventsManager.js';
import OptionsManager from './managers/OptionsManager.js';

let id = 0;

// eslint-disable-next-line no-undef
const isDev = typeof __DEV__ !== 'undefined' && __DEV__;

/**
 * No operation function.
 * @returns {void}
 */
function noop() {}

/**
 * Test if the managers' instances implement the default manager.
 *
 * @throws
 * @param   {Base} instance The instance to test.
 * @returns {void}
 */
function createAndTestManagers(instance) {
  [
    {
      prop: '__options',
      constructorName: 'OptionsManager',
      constructor: OptionsManager,
    },
    {
      prop: '__services',
      constructorName: 'ServicesManager',
      constructor: ServicesManager,
    },
    {
      prop: '__events',
      constructorName: 'EventsManager',
      constructor: EventsManager,
    },
    {
      prop: '__refs',
      constructorName: 'RefsManager',
      constructor: RefsManager,
    },
    {
      prop: '__children',
      constructorName: 'ChildrenManager',
      constructor: ChildrenManager,
    },
  ].forEach(({ prop, constructorName, constructor }) => {
    instance[prop] = new instance.__managers[constructorName](instance);
    if (!(instance[prop] instanceof constructor)) {
      throw new Error(
        `The \`$managers.${constructorName}\` must extend the \`${constructorName}\` class.`
      );
    }
  });
}

/**
 * @typedef {typeof Base} BaseConstructor
 * @typedef {(Base) => Promise<BaseConstructor | { default: BaseConstructor }>} BaseAsyncConstructor
 * @typedef {OptionsManager & { [name:string]: any }} BaseOptions
 * @typedef {{ [name:string]: HTMLElement | HTMLElement[] }} BaseRefs
 * @typedef {{ [nameOrSelector:string]: Base[] | Promise<Base>[] }} BaseChildren
 * @typedef {{ [nameOrSelector:string]: BaseConstructor | BaseAsyncConstructor }} BaseConfigComponents
 * @typedef {import('./managers/OptionsManager').OptionsSchema} BaseConfigOptions
 * @typedef {import('./managers/ServicesManager.js').ServiceName} ServiceName
 */

/**
 * @typedef {Object} BaseConfig
 * @property {string} name
 * @property {boolean} [debug]
 * @property {boolean} [log]
 * @property {string[]} [refs]
 * @property {string[]} [emits]
 * @property {BaseConfigComponents} [components]
 * @property {BaseConfigOptions} [options]
 */

/**
 * @typedef {Object} Managers
 * @property {typeof ChildrenManager} ChildrenManager
 * @property {typeof EventsManager} EventsManager
 * @property {typeof OptionsManager} OptionsManager
 * @property {typeof RefsManager} RefsManager
 * @property {typeof ServicesManager} ServicesManager
 */

/**
 * Base class.
 */
export default class Base extends EventTarget {
  /**
   * This is a Base instance.
   * @type {boolean}
   * @readonly
   */
  static $isBase = true;

  /**
   * The instance parent.
   * @type {Base}
   */
  $parent = null;

  /**
   * The instance id.
   * @type {string}
   */
  $id;

  /**
   * The root element.
   * @type {HTMLElement & { __base__?: WeakMap<BaseConstructor, Base | 'terminated'> }}
   */
  $el;

  /**
   * The state of the component.
   * @type {boolean}
   */
  $isMounted = false;

  /**
   * Store the event handlers.
   *
   * @private
   * @type {Map<string, Set<EventListenerOrEventListenerObject>>}
   */
  __eventHandlers = new Map();

  /**
   * Get the root instance of the app.
   * @returns {Base}
   */
  get $root() {
    if (!this.$parent) {
      return this;
    }

    let parent = this.$parent;
    let root = this.$parent;

    while (parent) {
      if (!parent.$parent) {
        root = parent;
      }

      parent = parent.$parent;
    }

    return root;
  }

  /**
   * Merge configuration with the parents' configurations.
   *
   * @returns {BaseConfig}
   * @private
   */
  get __config() {
    let proto = Object.getPrototypeOf(this);
    let { config } = proto.constructor;

    while (proto.constructor.config && proto.constructor.$isBase) {
      config = { ...proto.constructor.config, ...config };

      if (proto.constructor.config.options) {
        config.options = { ...proto.constructor.config.options, ...config.options };
      }

      if (proto.constructor.config.emits && config.emits) {
        config.emits = [...proto.constructor.config.emits, ...config.emits];
      }

      proto = Object.getPrototypeOf(proto);
    }

    config.options = config.options ?? {};
    config.refs = config.refs ?? [];
    config.components = config.components ?? {};

    return config;
  }

  /**
   * @type {BaseConfig}
   */
  static config = {
    name: 'Base',
    emits: [
      // hook events
      'mounted',
      'updated',
      'destroyed',
      'terminated',
      // default services' events
      'ticked',
      'scrolled',
      'resized',
      'moved',
      'loaded',
      'keyed',
    ],
  };

  /**
   * @type {ServicesManager}
   * @private
   */
  __services;

  /**
   * @type {ServicesManager}
   */
  get $services() {
    return this.__services;
  }

  /**
   * @type {RefsManager}
   * @private
   */
  __refs;

  /**
   * @returns {RefsManager}
   */
  get $refs() {
    return this.__refs;
  }

  /**
   * @type {BaseOptions}
   * @private
   */
  __options;

  /**
   * @returns {BaseOptions}
   */
  get $options() {
    return this.__options;
  }

  /**
   * @type {ChildrenManager}
   * @private
   */
  __children;

  /**
   * @returns {ChildrenManager}
   */
  get $children() {
    return this.__children;
  }

  /**
   * @type {EventsManager}
   * @private
   */
  __events;

  /**
   * Small helper to log stuff.
   *
   * @returns {(...args: any) => void}
   */
  get $log() {
    return this.__options.log ? window.console.log.bind(window, `[${this.__config.name}]`) : noop;
  }

  /**
   * Small helper to debug information.
   *
   * @private
   * @returns {(...args:any) => void}
   */
  get __debug() {
    return isDev && this.__options.debug
      ? window.console.log.bind(window, `[debug] [${this.$id}]`)
      : noop;
  }

  /**
   * Get manager constructors.
   *
   * @returns {Managers}
   */
  get __managers() {
    return {
      ChildrenManager,
      EventsManager,
      OptionsManager,
      RefsManager,
      ServicesManager,
    };
  }

  /**
   * Call an instance method and emit corresponding events.
   *
   * @private
   * @param {string} method
   * @param {any[]} args
   * @returns {void}
   */
  __callMethod(method, ...args) {
    if (isDev) {
      this.__debug('callMethod', method, ...args);
    }

    this.$emit(method, ...args);

    // We always emit an event, but we do not call the method if it does not exist
    if (typeof this[method] !== 'function') {
      return;
    }

    this[method].call(this, ...args);
    if (isDev) {
      this.__debug(method, this, ...args);
    }
  }

  /**
   * Test if the given event has been bound to the instance.
   *
   * @private
   * @param  {string} event The event's name.
   * @returns {boolean}      Wether the given event has been bound or not.
   */
  __hasEvent(event) {
    const eventHandlers = this.__eventHandlers.get(event);
    return eventHandlers && eventHandlers.size > 0;
  }

  /**
   * Class constructor where all the magic takes place.
   *
   * @param {HTMLElement} element The component's root element dd.
   */
  constructor(element) {
    super();

    if (!element) {
      throw new Error('The root element must be defined.');
    }

    const { __config } = this;

    if (__config.name === 'Base') {
      throw new Error('The `config.name` property is required.');
    }

    this.$id = `${__config.name}-${id}`;
    id += 1;

    this.$el = element;

    if (!this.$el.__base__) {
      this.$el.__base__ = new WeakMap();
    }

    this.$el.__base__.set(this.__ctor, this);

    createAndTestManagers(this);

    if (isDev) {
      this.__debug('constructor', this);
    }
  }

  /**
   * Trigger the `mounted` callback.
   * @returns {this}
   */
  $mount() {
    if (this.$isMounted) {
      return this;
    }

    this.$isMounted = true;
    if (isDev) {
      this.__debug('$mount');
    }
    this.$children.registerAll();
    this.$refs.registerAll();
    this.__events.bindRootElement();
    this.$services.enableAll();
    this.$children.mountAll();

    this.__callMethod('mounted');

    return this;
  }

  /**
   * Update the instance children.
   * @returns {this}
   */
  $update() {
    if (isDev) {
      this.__debug('$update');
    }

    // Undo
    this.$refs.unregisterAll();
    this.$services.disableAll();

    // Redo
    this.$children.registerAll();
    this.$refs.registerAll();
    this.$services.enableAll();

    // Update
    this.$children.updateAll();

    this.__callMethod('updated');
    return this;
  }

  /**
   * Trigger the `destroyed` callback.
   * @returns {this}
   */
  $destroy() {
    if (!this.$isMounted) {
      return this;
    }

    this.$isMounted = false;
    if (isDev) {
      this.__debug('$destroy');
    }
    this.__events.unbindRootElement();
    this.$refs.unregisterAll();
    this.$services.disableAll();
    this.$children.destroyAll();
    this.__callMethod('destroyed');

    return this;
  }

  /**
   * Terminate a child instance when it is not needed anymore.
   * @returns {void}
   */
  $terminate() {
    if (isDev) {
      this.__debug('$terminate');
    }

    // First, destroy the component.
    this.$destroy();

    // Execute the `terminated` hook if it exists
    this.__callMethod('terminated');

    // Delete instance
    this.$el.__base__.set(this.__ctor, 'terminated');
  }

  /**
   * Test if the given event is configured and warn otherwise.
   *
   * @private
   * @param   {string} event The event name.
   * @returns {void}
   */
  __testEventIsDefined(event) {
    if (!(Array.isArray(this.__config.emits) && this.__config.emits.includes(event))) {
      console.warn(
        `[${this.__config.name}]`,
        `The "${event}" event is missing from the configuration.`
      );
    }
  }

  /**
   * Add an emitted event.
   *
   * @private
   * @param   {string} event The event name.
   * @returns {void}
   */
  __addEmits(event) {
    const ctor = this.__ctor;
    if (Array.isArray(ctor.config.emits)) {
      ctor.config.emits.push(event);
    } else {
      ctor.config.emits = [event];
    }
  }

  /**
   * Remove an emitted event.
   *
   * @private
   * @param   {string} event The event name.
   * @returns {void}
   */
  __removeEmits(event) {
    const ctor = this.__ctor;
    const index = ctor.config.emits.indexOf(event);
    ctor.config.emits.splice(index, 1);
  }

  /**
   * Get the instance constructor.
   *
   * @private
   * @returns {BaseConstructor}
   */
  get __ctor() {
    return /** @type {BaseConstructor} */ (this.constructor);
  }

  /**
   * Bind a listener function to an event.
   *
   * @param  {string} event
   *   Name of the event.
   * @param  {EventListenerOrEventListenerObject} listener
   *   Function to be called.
   * @param {boolean|AddEventListenerOptions} [options]
   *   Options for the `removeEventListener` method.
   * @returns {() => void}
   *   A function to unbind the listener.
   */
  $on(event, listener, options) {
    if (isDev) {
      this.__debug('$on', event, listener, options);
    }

    if (isDev) {
      this.__testEventIsDefined(event);
    }

    let set = this.__eventHandlers.get(event);

    if (!set) {
      set = new Set();
      this.__eventHandlers.set(event, set);
    }

    set.add(listener);
    this.addEventListener(event, listener, options);

    return () => {
      this.$off(event, listener, options);
    };
  }

  /**
   * Unbind a listener function from an event.
   *
   * @param {string} event
   *   Name of the event.
   * @param {EventListenerOrEventListenerObject} listener
   *   Function to be removed.
   * @param {boolean|EventListenerOptions} [options]
   *   Options for the `removeEventListener` method.
   * @returns {void}
   */
  $off(event, listener, options) {
    if (isDev) {
      this.__debug('$off', event, listener);
    }

    if (isDev) {
      this.__testEventIsDefined(event);
    }

    this.__eventHandlers.get(event).delete(listener);
    this.removeEventListener(event, listener, options);
  }

  /**
   * Emits an event.
   *
   * @param  {string} event
   *   Name of the event.
   * @param  {any[]}        args  The arguments to apply to the functions bound to this event.
   * @returns {void}
   */
  $emit(event, ...args) {
    if (isDev) {
      this.__debug('$emit', event, args);
    }

    if (isDev) {
      this.__testEventIsDefined(event);
    }

    this.dispatchEvent(new CustomEvent(event, { detail: args }));
  }

  /**
   * Factory method to generate multiple instance of the class.
   *
   * @param  {string}      nameOrSelector The selector on which to mount each instance.
   * @returns {Array<Base>}                A list of the created instance.
   */
  static $factory(nameOrSelector) {
    if (!nameOrSelector) {
      throw new Error(
        'The $factory method requires a component’s name or selector to be specified.'
      );
    }

    return getComponentElements(nameOrSelector).map((el) => new this(el).$mount());
  }
}
