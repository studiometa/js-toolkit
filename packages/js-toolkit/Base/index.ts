/* eslint-disable no-use-before-define */
import { getComponentElements, getEventTarget } from './utils.js';
import { ChildrenManager } from './managers/ChildrenManager.js';
import { RefsManager } from './managers/RefsManager.js';
import { ServicesManager } from './managers/ServicesManager.js';
import { EventsManager } from './managers/EventsManager.js';
import { OptionsManager } from './managers/OptionsManager.js';
import { noop, isDev, isFunction, isArray } from '../utils/index.js';

let id = 0;

/**
 * Test if the managers' instances implement the default manager.
 * @throws
 */
function createAndTestManagers(instance: Base): void {
  [
    {
      prop: '__options',
      constructorName: 'OptionsManager',
      constructor: OptionsManager,
    },
    {
      prop: '__services',
      constructorName: 'ServicesManager',
      constructor: ServicesManager,
    },
    {
      prop: '__events',
      constructorName: 'EventsManager',
      constructor: EventsManager,
    },
    {
      prop: '__refs',
      constructorName: 'RefsManager',
      constructor: RefsManager,
    },
    {
      prop: '__children',
      constructorName: 'ChildrenManager',
      constructor: ChildrenManager,
    },
  ].forEach(({ prop, constructorName, constructor }) => {
    instance[prop] = new instance.__managers[constructorName](instance);
    if (isDev && !(instance[prop] instanceof constructor)) {
      throw new Error(
        `The \`$managers.${constructorName}\` must extend the \`${constructorName}\` class.`,
      );
    }
  });
}

export type BaseEl = HTMLElement & { __base__?: WeakMap<BaseConstructor, Base | 'terminated'> };
export type BaseConstructor<T extends Base = Base> = {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  new (...args: any[]): T;
  prototype: Base;
} & Pick<typeof Base, keyof typeof Base>;
export type BaseAsyncConstructor<T extends Base = Base> = (
  Base
) => Promise<BaseConstructor<T> | { default: BaseConstructor<T> }>;
export type BaseOptions = { [name: string]: unknown };
export type BaseRefs = { [name: string]: HTMLElement | HTMLElement[] };
export type BaseChildren = { [nameOrSelector: string]: Base[] | Promise<Base>[] };
export type BaseConfigComponents = {
  [nameOrSelector: string]: BaseConstructor | BaseAsyncConstructor;
};

export type BaseConfig = {
  name: string;
  debug?: boolean;
  log?: boolean;
  refs?: string[];
  emits?: string[];
  components?: BaseConfigComponents;
  options?: import('./managers/OptionsManager').OptionsSchema;
};

export type BaseProps = {
  $el?: HTMLElement;
  $options?: BaseOptions;
  $refs?: BaseRefs;
  $children?: BaseChildren;
  $parent?: Base;
};

export type Managers = {
  ChildrenManager: typeof ChildrenManager;
  EventsManager: typeof EventsManager;
  OptionsManager: typeof OptionsManager;
  RefsManager: typeof RefsManager;
  ServicesManager: typeof ServicesManager;
};

/**
 * Base class.
 */
export class Base<T extends BaseProps = BaseProps> extends EventTarget {
  /**
   * This is a Base instance.
   */
  static readonly $isBase = true as const;

  /**
   * The instance parent.
   */
  $parent: T['$parent'] & Base = null;

  /**
   * The instance id.
   */
  $id: string;

  /**
   * The root element.
   */
  $el: T['$el'] & BaseEl;

  /**
   * The state of the component.
   */
  $isMounted = false;

  /**
   * Store the event handlers.
   */
  __eventHandlers: Map<string, Set<EventListenerOrEventListenerObject>> = new Map();

  /**
   * Get the root instance of the app.
   */
  get $root(): Base {
    if (!this.$parent) {
      return this;
    }

    let parent = this.$parent;
    let root = this.$parent;

    while (parent) {
      if (!parent.$parent) {
        root = parent;
      }

      parent = parent.$parent;
    }

    return root;
  }

  /**
   * Merge configuration with the parents' configurations.
   */
  get __config(): BaseConfig {
    let proto = Object.getPrototypeOf(this);
    let { config } = proto.constructor;

    while (proto.constructor.config && proto.constructor.$isBase) {
      config = { ...proto.constructor.config, ...config };

      if (proto.constructor.config.options) {
        config.options = { ...proto.constructor.config.options, ...config.options };
      }

      if (proto.constructor.config.emits && config.emits) {
        config.emits = [...proto.constructor.config.emits, ...config.emits];
      }

      proto = Object.getPrototypeOf(proto);
    }

    config.options = config.options ?? {};
    config.refs = config.refs ?? [];
    config.components = config.components ?? {};

    return config;
  }

  static config: BaseConfig = {
    name: 'Base',
    emits: [
      // hook events
      'before-mounted',
      'mounted',
      'updated',
      'destroyed',
      'terminated',
      // default services' events
      'ticked',
      'scrolled',
      'resized',
      'moved',
      'loaded',
      'keyed',
    ],
  };

  __services: ServicesManager;

  get $services(): ServicesManager {
    return this.__services;
  }

  __refs: RefsManager & T['$refs'] & BaseRefs;

  get $refs(): RefsManager & T['$refs'] & BaseRefs {
    return this.__refs;
  }

  __options: OptionsManager & T['$options'] & BaseOptions;

  get $options(): OptionsManager & T['$options'] & BaseOptions {
    return this.__options;
  }

  __children: ChildrenManager & T['$children'] & BaseChildren;

  get $children(): ChildrenManager & T['$children'] & BaseChildren {
    return this.__children;
  }

  __events: EventsManager;

  /**
   * Small helper to log stuff.
   */
  get $log(): (...args: unknown[]) => void {
    return this.__options.log ? window.console.log.bind(window, `[${this.__config.name}]`) : noop;
  }

  /**
   * Small helper to debug information.
   */
  get __debug(): (...args: unknown[]) => void {
    return isDev && this.__options.debug
      ? window.console.log.bind(window, `[debug] [${this.$id}]`)
      : noop;
  }

  /**
   * Get manager constructors.
   */
  get __managers(): Managers {
    return {
      ChildrenManager,
      EventsManager,
      OptionsManager,
      RefsManager,
      ServicesManager,
    };
  }

  /**
   * Call an instance method and emit corresponding events.
   */
  __callMethod(method: string, ...args: unknown[]): unknown {
    if (isDev) {
      this.__debug('callMethod', method, ...args);
    }

    this.$emit(method, ...args);

    // We always emit an event, but we do not call the method if it does not exist
    if (!isFunction(this[method])) {
      return null;
    }

    if (isDev) {
      this.__debug(method, this, ...args);
    }

    return this[method].call(this, ...args);
  }

  /**
   * Test if the given event has been bound to the instance.
   *
   * @param   {string} event The event's name.
   * @returns {boolean}      Wether the given event has been bound or not.
   */
  __hasEvent(event: string): boolean {
    const eventHandlers = this.__eventHandlers.get(event);
    return eventHandlers && eventHandlers.size > 0;
  }

  /**
   * Class constructor where all the magic takes place.
   *
   * @param {HTMLElement} element The component's root element dd.
   */
  constructor(element: HTMLElement) {
    super();

    if (!element) {
      if (isDev) {
        throw new Error('The root element must be defined.');
      }
      return;
    }

    const { __config } = this;

    if (__config.name === 'Base') {
      if (isDev) {
        throw new Error('The `config.name` property is required.');
      }
      return;
    }

    this.$id = `${__config.name}-${id}`;
    id += 1;

    this.$el = element;

    if (!this.$el.__base__) {
      this.$el.__base__ = new WeakMap();
    }

    this.$el.__base__.set(this.__ctor, this);

    createAndTestManagers(this);

    if (isDev) {
      this.__debug('constructor', this);
    }
  }

  /**
   * Trigger the `mounted` callback.
   */
  $mount(): this {
    if (this.$isMounted) {
      return this;
    }

    this.$emit('before-mounted');

    this.$isMounted = true;
    if (isDev) {
      this.__debug('$mount');
    }
    this.$children.registerAll();
    this.$refs.registerAll();
    this.__events.bindRootElement();
    this.$services.enableAll();
    this.$children.mountAll();

    this.__callMethod('mounted');

    return this;
  }

  /**
   * Update the instance children.
   */
  $update(): this {
    if (isDev) {
      this.__debug('$update');
    }

    // Undo
    this.$refs.unregisterAll();
    this.$services.disableAll();

    // Redo
    this.$children.registerAll();
    this.$refs.registerAll();
    this.$services.enableAll();

    // Update
    this.$children.updateAll();

    this.__callMethod('updated');
    return this;
  }

  /**
   * Trigger the `destroyed` callback.
   */
  $destroy(): this {
    if (!this.$isMounted) {
      return this;
    }

    this.$isMounted = false;
    if (isDev) {
      this.__debug('$destroy');
    }
    this.__events.unbindRootElement();
    this.$refs.unregisterAll();
    this.$services.disableAll();
    this.$children.destroyAll();
    this.__callMethod('destroyed');

    return this;
  }

  /**
   * Terminate a child instance when it is not needed anymore.
   */
  $terminate(): void {
    if (isDev) {
      this.__debug('$terminate');
    }

    // First, destroy the component.
    this.$destroy();

    // Execute the `terminated` hook if it exists
    this.__callMethod('terminated');

    // Delete instance
    this.$el.__base__.set(this.__ctor, 'terminated');
  }

  /**
   * Add an emitted event.
   *
   * @param   {string} event The event name.
   * @returns {void}
   */
  __addEmits(event: string): void {
    const ctor = this.__ctor;
    if (isArray(ctor.config.emits)) {
      ctor.config.emits.push(event);
    } else {
      ctor.config.emits = [event];
    }
  }

  /**
   * Remove an emitted event.
   *
   * @param   {string} event The event name.
   * @returns {void}
   */
  __removeEmits(event: string): void {
    const ctor = this.__ctor;
    const index = ctor.config.emits.indexOf(event);
    ctor.config.emits.splice(index, 1);
  }

  /**
   * Get the instance constructor.
   */
  get __ctor(): BaseConstructor {
    return this.constructor as BaseConstructor;
  }

  /**
   * Bind a listener function to an event.
   *
   * @param  {string} event
   *   Name of the event.
   * @param  {EventListenerOrEventListenerObject} listener
   *   Function to be called.
   * @param {boolean|AddEventListenerOptions} [options]
   *   Options for the `removeEventListener` method.
   * @returns {() => void}
   *   A function to unbind the listener.
   */
  $on(
    event: string,
    listener: EventListenerOrEventListenerObject,
    options?: boolean | AddEventListenerOptions,
  ): () => void {
    if (isDev) {
      this.__debug('$on', event, listener, options);
    }

    let set = this.__eventHandlers.get(event);

    if (!set) {
      set = new Set();
      this.__eventHandlers.set(event, set);
    }

    set.add(listener);

    const target = getEventTarget(this, event, this.__config);
    target.addEventListener(event, listener, options);

    return () => {
      this.$off(event, listener, options);
    };
  }

  /**
   * Unbind a listener function from an event.
   *
   * @param {string} event
   *   Name of the event.
   * @param {EventListenerOrEventListenerObject} listener
   *   Function to be removed.
   * @param {boolean|EventListenerOptions} [options]
   *   Options for the `removeEventListener` method.
   * @returns {void}
   */
  $off(
    event: string,
    listener: EventListenerOrEventListenerObject,
    options?: boolean | EventListenerOptions,
  ) {
    if (isDev) {
      this.__debug('$off', event, listener);
    }

    this.__eventHandlers.get(event).delete(listener);

    const target = getEventTarget(this, event, this.__config);
    target.removeEventListener(event, listener, options);
  }

  /**
   * Emits an event.
   *
   * @param  {string} event Name of the event.
   * @param  {any[]}  args  The arguments to apply to the functions bound to this event.
   * @returns {void}
   */
  $emit(event: string, ...args: unknown[]) {
    if (isDev) {
      this.__debug('$emit', event, args);
    }

    this.dispatchEvent(new CustomEvent(event, { detail: args }));
  }

  /**
   * Factory method to generate multiple instance of the class.
   */
  static $factory(nameOrSelector: string): Base[] {
    if (isDev && !nameOrSelector) {
      throw new Error(
        'The $factory method requires a component’s name or selector to be specified.',
      );
    }

    return getComponentElements(nameOrSelector).map((el) => new this(el).$mount());
  }
}
