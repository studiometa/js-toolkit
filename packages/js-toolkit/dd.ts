/* eslint-disable */
interface B {
  $children?: Record<string, Base[]>;
}

class Base<T extends B = B> {
  $children: T['$children'];
}

class Foo extends Base {
  constructor() {
    super();
    this.$children;
  }
}

class Bar<O extends B = B> extends Base<{ $children: { Foo: Foo[] } }> {
  constructor() {
    super();
    this.$children.Foo;
  }

  async enter() {}
}

interface WithDecoratorInterface extends B {
  $children: {
    Bar: Bar[];
  };
}

type Constructor<T, Arguments extends unknown[] = any[]> = new (...arguments_: Arguments) => T;

function withDecorator<T extends Base, V extends B = B>(BaseClass: Constructor<T>) {
  class WithDecorator<W extends V = V> extends BaseClass<WithDecoratorInterface & W & V> {
    /**
     * Dede.
     */
    dede() {}
  }

  return WithDecorator as Constructor<T> & {
    new <W extends B = B>(): T &
      WithDecorator &
      Base<W & V & WithDecoratorInterface>;
  };
}


class Baz extends withDecorator<Bar, { $children: { Foo: Foo[] } }>(Bar) {
  constructor() {
    super();
    this.$children.Foo;
    this.$children.Bar[0].enter();
    this.enter();
    this.dede();
  }
}

const BozTemp = withDecorator(Baz);

/**
 * @extends {BozTemp<{ $children: { Dede: Foo[] } }> }
 */
class Boz extends BozTemp {
  constructor() {
    super();
    this.$children.Foo;
    this.dede();
    this.$children.Dede;
  }
}
