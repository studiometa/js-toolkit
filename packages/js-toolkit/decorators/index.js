import withBreakpointManager from './withBreakpointManager.js';
import withBreakpointObserver from './withBreakpointObserver.js';
import withExtraConfig from './withExtraConfig.js';
import withIntersectionObserver from './withIntersectionObserver.js';
import withMountWhenInView from './withMountWhenInView.js';
import withVue from './withVue.js';

export {
  withBreakpointManager,
  withBreakpointObserver,
  withExtraConfig,
  withIntersectionObserver,
  withMountWhenInView,
  withVue,
};
