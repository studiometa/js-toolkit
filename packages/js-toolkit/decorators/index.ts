export * from './withBreakpointManager.js';
export * from './withBreakpointObserver.js';
export * from './withDrag.js';
export * from './withExtraConfig.js';
export * from './withFreezedOptions.js';
export * from './withIntersectionObserver.js';
export * from './withMountOnMediaQuery.js';
export * from './withMountWhenInView.js';
export * from './withMountWhenPrefersMotion.js';
export * from './withMutation.js';
export * from './withName.js';
export * from './withRelativePointer.js';
export * from './withResponsiveOptions.js';
export * from './withScrolledInView/index.js';
export * from './withWheel.js';
