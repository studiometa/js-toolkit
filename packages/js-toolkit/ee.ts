/* eslint-disable */
declare type Constructor<T> = new (...args: any[]) => T;

type BaseChildren = Record<string, Base[]>;

type BaseType = {
  $children?: BaseChildren;
};

class Base<T extends BaseType = BaseType> {
  mounted() {
    return true;
  }

  static isBase = true as const;

  $children: T['$children'] & BaseChildren = {};
}

type MixinInterface = {
  $children: {
    Doo: Dede[];
  };
};

function withType<T extends Constructor<Base>, U extends BaseType = BaseType>(superclass: T) {
  return superclass as T &
    Constructor<Base<U>> &
    Pick<typeof Base, keyof typeof Base>;
}

function withDecorator<T extends Constructor<Base>, U extends BaseType = BaseType>(superclass: T) {
  class MixinClass extends superclass {
    mixin(): number {
      this.mounted();
      return 10;
    }
  }

  return MixinClass as Constructor<MixinClass> &
    Pick<typeof MixinClass, keyof typeof MixinClass> &
    T &
    Constructor<Base<U & MixinInterface>> &
    Pick<typeof Base, keyof typeof Base>;
}

class Dede extends Base {
  dodo() {
    return false as const;
  }
}

class Foo extends withDecorator<typeof Base, { $children: { Foo: Dede[] } }>(Base) {
  foo() {
    return 'foo';
  }

  constructor() {
    super();
    this.mounted();
    this.$children.Foo[0].dodo();
  }
}

class Bar extends withDecorator<typeof Foo, { $children: { Poo: Foo[] } }>(Foo) {
  name = 'Bar';
  constructor() {
    super();
    this.$children;
    this.mounted();
    this.foo();
    this.mixin();
    this.$children.Poo[0].mounted();
    this.$children.Foo[0].dodo;
  }
}

// @todo find a way to make Bar accept a type parameter
// maybe following https://ybogomolov.me/01-higher-kinded-types/

class Boz extends withType<typeof Bar, { $children: { Boz: Bar[] } }>(Bar) {
  name = 'Boz';

  constructor() {
    super();
    this.name;
    this.$children;
    this.mounted();
    this.foo();
    this.mixin();
    this.$children.Poo[0].mounted();
    this.$children.Foo[0].dodo;
    this.$children.Boz[0].name;
  }
}


class One extends Base<{ $children: { One: One[] } }> {
  mounted() {
    this.$children.One[0].mounted;
    return true;
  }
}
