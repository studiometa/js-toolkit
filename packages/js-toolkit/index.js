export { default as Base } from './Base/index.js';
export * from './decorators/index.js';
export * from './helpers/index.js';
export * from './services/index.js';
