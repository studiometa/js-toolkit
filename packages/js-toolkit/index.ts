export * from './Base/index.js';
export * from './Base/types.js';
export * from './decorators/index.js';
export * from './helpers/index.js';
export * from './services/index.js';
