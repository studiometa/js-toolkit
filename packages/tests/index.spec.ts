import * as toolkit from '@studiometa/js-toolkit';

describe('The package exports', () => {
  it('should export helpers and the Base class', () => {
    expect(Object.keys(toolkit)).toMatchSnapshot();
  });
});
