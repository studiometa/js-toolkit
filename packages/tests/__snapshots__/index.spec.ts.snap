// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`The package exports should export helpers and the Base class 1`] = `
[
  "Base",
  "createApp",
  "getClosestParent",
  "getDirectChildren",
  "getInstanceFromElement",
  "getInstances",
  "importOnInteraction",
  "importOnMediaQuery",
  "importWhenIdle",
  "importWhenPrefersMotion",
  "importWhenVisible",
  "isDirectChild",
  "useDrag",
  "useKey",
  "useLoad",
  "usePointer",
  "useRaf",
  "useResize",
  "useScroll",
  "useService",
  "withBreakpointManager",
  "withBreakpointObserver",
  "withDrag",
  "withExtraConfig",
  "withFreezedOptions",
  "withIntersectionObserver",
  "withMountOnMediaQuery",
  "withMountWhenInView",
  "withMountWhenPrefersMotion",
  "withName",
  "withRelativePointer",
  "withResponsiveOptions",
  "withScrolledInView",
]
`;
